src/AstTransforms.cpp

```diff
@@ -220,37 +221,66 @@ bool MaterializeAggregationQueriesTransformer::materializeAggregationQueries(
     // if an aggregator has a body consisting of more than an atom => create new relation
     int counter = 0;
     visitDepthFirst(program, [&](const AstClause& clause) {
-        visitDepthFirstPostOrder(clause, [&](const AstAggregator& agg) {
+        visitDepthFirst(clause, [&](const AstAggregator& agg) {
             // check whether a materialization is required
             if (!needsMaterializedRelation(agg)) {
                 return;
             }
-
             changed = true;
 
-            // for more body literals: create a new relation
-            std::set<std::string> vars;
-            visitDepthFirst(agg, [&](const AstVariable& var) { vars.insert(var.getName()); });
-
             // -- create a new clause --
 
             auto relName = "__agg_rel_" + toString(counter++);
             while (program.getRelation(relName) != nullptr) {
                 relName = "__agg_rel_" + toString(counter++);
             }
+            // create the new clause for the materialised thing
+            auto* aggClause = new AstClause();
+            // create the body of the new thing
+            for (const auto& cur : agg.getBodyLiterals()) {
+                aggClause->addToBody(std::unique_ptr<AstLiteral>(cur->clone()));
+            }
+            // find stuff for which we need a grounding
+            for (const auto& argPair : getGroundedTerms(*aggClause)) {
+              const AstVariable* variable = dynamic_cast<const AstVariable*>(argPair.first);
```

Comment:
Use auto rather than duplicating type.
```suggestion
              const auto* variable = dynamic_cast<const AstVariable*>(argPair.first);
```